explorer/src/Pos/Explorer/Web/Server.hs

```diff
@@ -522,34 +527,77 @@ getRedeemAddressCoinPairs = do
 
     pure redeemOnly
 
-isAddressRedeemed :: MonadDBRead m => Address -> Coin -> m Bool
-isAddressRedeemed address initialBalance = do
-  currentBalance <- fromMaybe (mkCoin 0) <$> EX.getAddrBalance address
-  pure $ currentBalance /= initialBalance
-
 getGenesisSummary
     :: ExplorerMode ctx m
     => m CGenesisSummary
 getGenesisSummary = do
     redeemAddressCoinPairs <- getRedeemAddressCoinPairs
-    cgsNumRedeemed <- length <$> filterM (uncurry isAddressRedeemed) redeemAddressCoinPairs
-    pure CGenesisSummary {cgsNumTotal = length redeemAddressCoinPairs, ..}
+    redeemAddressInfo <- mapM (uncurry getRedeemAddressInfo) redeemAddressCoinPairs
+    let (cgsNumRedeemed, redeemedAmountTotal, nonRedeemedAmountTotal) =
+            foldr folder (0, mkCoin 0, mkCoin 0) redeemAddressInfo
+    let cgsNumTotal = length redeemAddressCoinPairs
+    pure CGenesisSummary
+        { cgsNumNotRedeemed = cgsNumTotal - cgsNumRedeemed
+        , cgsRedeemedAmountTotal = mkCCoin redeemedAmountTotal
+        , cgsNonRedeemedAmountTotal = mkCCoin nonRedeemedAmountTotal
+        , ..
+        }
+  where
+    getRedeemAddressInfo
+        :: (MonadDBRead m, MonadThrow m)
+        => Address -> Coin -> m (Bool, Coin, Coin)
+    getRedeemAddressInfo address initialBalance = do
+        currentBalance <- fromMaybe (mkCoin 0) <$> EX.getAddrBalance address
+        if currentBalance > initialBalance then
+            throwM $ Internal $ sformat
+                ("Redeem address "%build%" had "%build%" at genesis, but now has "%build)
+                address initialBalance currentBalance
+        else
+            let isRedeemed = currentBalance == mkCoin 0
+                redeemedAmount = initialBalance `unsafeSubCoin` currentBalance
+                amountLeft = currentBalance
+            in pure $ (isRedeemed, redeemedAmount, amountLeft)
+    folder :: (Bool, Coin, Coin) -> (Int, Coin, Coin) -> (Int, Coin, Coin)
```

Comment:
Same here. It's optional, but this could be more readable and less prone to error.